internal/factory/container/mounts.go

```diff
@@ -0,0 +1,285 @@
+package container
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/containers/storage/pkg/idtools"
+	"github.com/containers/storage/pkg/mount"
+	"github.com/cri-o/cri-o/internal/lib/sandbox"
+	oci "github.com/cri-o/cri-o/internal/oci"
+	"github.com/cri-o/cri-o/internal/resourcestore"
+	"github.com/cri-o/cri-o/internal/storage"
+	sconfig "github.com/cri-o/cri-o/pkg/config"
+	securejoin "github.com/cyphar/filepath-securejoin"
+	rspec "github.com/opencontainers/runtime-spec/specs-go"
+	types "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+type mountInfo struct {
+	criMounts map[string]*rspec.Mount
+	criDevSet bool
+	criSysSet bool
+	mounts    map[string]*rspec.Mount
+}
+
+type orderedMounts []*rspec.Mount
+
+// SpecAddPreOCIMounts add mounts to the spec before creating ocicontainer
+func (c *container) SpecAddPreOCIMounts(ctx context.Context, resourceStore *resourcestore.ResourceStore, serverConfig *sconfig.Config, sb *sandbox.Sandbox, containerInfo storage.ContainerInfo, mountPoint string, idMapSupport bool) ([]oci.ContainerVolume, []rspec.Mount, error) {
+	// Create temp mountInfo
+	c.mountInfo = newMountInfo()
+
+	// Clear temp mountInfo
+	defer clearMountInfo(c)
```

Comment:
if we clear mountinfo between these two calls (pre and post) then we lose the deduplication right?